Guide to properly uninstall submariner from a cluster (#411)
submariner-io/submariner-website: src/content/operations/cleanup/_index.en.md

---
title: "Uninstalling Submariner"
date: 2020-12-23T21:25:11+01:00
weight: 50
---

> To properly uninstall Submariner from a cluster, follow the steps below:

Reviewer: ```suggestion
To properly uninstall Submariner from a cluster, follow the steps below.
```

Or

```suggestion
To properly uninstall Submariner from a cluster:
```

> 1. Delete Submariner related namespaces

Reviewer: ```suggestion
1. Delete Submariner-related namespaces
```

>    On each participating cluster, issue the following command:

Reviewer: ```suggestion
   On each participating cluster:
```

Maybe remove this repetitive line from each section?

Author: I kind of like it for the sort of descriptiveness.

   ```bash
   kubectl delete namespace submariner-operator
   ```

   On the Broker cluster, issue the following command:

   ```bash
   kubectl delete namespace submariner-k8s-broker
    ```

> 2. Delete submariner CRDs

Reviewer 2: ```suggestion
2. Delete the Submariner CRDs
```

   On each participating cluster, issue the following command:

   ```bash
   for CRD in `kubectl get crds | grep -iE 'submariner|multicluster.x-k8s.io'| awk '{print $1}'`; do kubectl delete crd $CRD; done
   ```

> 3. Delete ClusterRoles and ClusterRoleBindings

Reviewer: ```suggestion
3. Delete Submariner's `ClusterRole`s and `ClusterRoleBinding`s
```

https://kubernetes.io/docs/contribute/style/style-guide/#use-code-style-for-inline-code-commands-and-api-objects

   On each participating cluster, issue the following command:

   ```bash
   roles="submariner-operator submariner-operator-globalnet submariner-lighthouse submariner-networkplugin-syncer"
   kubectl delete clusterrole,clusterrolebinding $roles --ignore-not-found
   ```

4. Remove Submariner's iptables chains

>    On all nodes in each participating cluster, issue the following command:

Reviewer 2: ```suggestion
   On all nodes in each participating cluster, issue the following commands:
```

   ```bash
   iptables --flush SUBMARINER-INPUT
   iptables -D INPUT $(iptables -L INPUT --line-numbers | grep SUBMARINER-INPUT | awk '{print $1}')
   iptables --delete-chain SUBMARINER-INPUT
   
   iptables --flush SUBMARINER-POSTROUTING
   iptables -D INPUT $(iptables -L INPUT --line-numbers | grep SUBMARINER-POSTROUTING | awk '{print $1}')
   iptables --delete-chain SUBMARINER-POSTROUTING
   ```

>    If Globalnet is enabled in the setup, delete below iptable chains as well

Reviewer 2: ```suggestion
   If Globalnet is enabled in the setup, issue the following commands as well:
```

   ```bash
   iptables --flush SUBMARINER-GN-INGRESS
   iptables -D INPUT $(iptables -L INPUT --line-numbers | grep SUBMARINER-GN-INGRESS | awk '{print $1}')
   iptables --delete-chain SUBMARINER-GN-INGRESS

   iptables --flush SUBMARINER-GN-EGRESS
   iptables -D INPUT $(iptables -L INPUT --line-numbers | grep SUBMARINER-GN-EGRESS | awk '{print $1}')
   iptables --delete-chain SUBMARINER-GN-EGRESS

   iptables --flush SUBMARINER-GN-MARK
   iptables -D INPUT $(iptables -L INPUT --line-numbers | grep SUBMARINER-GN-MARK | awk '{print $1}')
   iptables --delete-chain SUBMARINER-GN-MARK
   ```

   {{% notice note %}}

>    Note that Kind based setup will have just `SUBMARINER-INPUT` chain.

Reviewer: ```suggestion
   Note that kind-based setup will have just `SUBMARINER-INPUT` chain.
```

https://submariner.io/development/website/style_guide/#submarinerio-word-list

>    Note that Kind based setup will have just `SUBMARINER-INPUT` chain.

Reviewer 3: @Jaanki I'm not sure if this statement (i.e., Note that Kind based setup will have just `SUBMARINER-INPUT` chain) is valid. Can you please share additional info on why KIND based setup will have just SUBMARINER-INPUT chain and not others?

Author: @sridhargaddam I am not seeing other chains in AWS setup also. Not even on gateway nodes. Just SUBMARINER-INPUT chain.

Reviewer 3: Probably you are looking at the wrong place. Please ping me when your setup is available and I'll walk-through how to look at various chains.

Author: ohk so the iptables are programmed inside the globalnet pod. I was looking on the nodes.

Reviewer 3: They are programmed on the node itself. All Submariner Pods (i.e., Globalnet, Gateway and RouteAgent) run with HostNetworking set to true. So when you exec to the pods and check the iptables configuration, you are actually looking at the configuration on the nodes itself.

   {{% /notice %}}

> 5. Delete `vx-submariner` interface

Reviewer 2: ```suggestion
5. Delete the `vx-submariner` interface
```

   On all nodes in each participating cluster, issue the following command:

   ```bash
   ip link delete vx-submariner
   ```

6. Remove the Submariner gateway labels

   On all nodes in each participating cluster, issue the following command:

   ```bash
   kubectl label --all node submariner.io/gateway-
   ```

> 7. For OpenShift deployments, delete `Lighthouse ServiceExport DNS`

Reviewer 4: ```suggestion
7. For OpenShift deployments, delete Lighthouse entry from `default` DNS.
```

> 7. For Openshift deployments, delete `Lighthouse ServiceExport DNS`

Reviewer 2: ```suggestion
7. For OpenShift deployments, delete `Lighthouse ServiceExport DNS`
```

> 7. For Openshift deployments, delete `Lighthouse ServiceExport DNS`

Reviewer 2: What do you mean by `Lighthouse ServiceExport DNS`? Isn't it the Lighthouse CoreDNS forward plugin?

   For each participating cluster, issue the following command:

   ```bash
   roles="submariner-operator submariner-operator-globalnet submariner-lighthouse submariner-networkplugin-syncer"
   kubectl delete clusterrole,clusterrolebinding $roles --ignore-not-found
   ```

4. Remove the Submariner gateway labels

   For each participating cluster, issue the following command:

   ```bash
   kubectl label --all node submariner.io/gateway-
   ```

5. For OpenShift deployments, delete Lighthouse entry from `default` DNS.

   For each participating cluster, issue the following command:

   ```bash
   kubectl apply -f -  <<EOF
   apiVersion: operator.openshift.io/v1
   kind: DNS
   metadata:
     finalizers:
     - dns.operator.openshift.io/dns-controller
     name: default
   spec:
     servers: []
   EOF
   ```

6. Delete Submariner-related configmaps

   For each participating cluster, issue the following commands:

   ```bash
   kubectl delete configmap submariner-engine-lock -n submariner-operator
   kubectl delete configmap submariner-operator-lock -n submariner-operator
   kubectl delete configmap submariner-lighthouse-coredns -n submariner-operator

>    ```

Reviewer 3: In Step 1, if we delete the `submariner-operator` namespace, all the configmaps will be deleted automatically.

Reviewer 2: Correct, this step 6 is not needed after the namespace is gone:

```
$ kubectl delete configmap submariner-engine-lock -n submariner-operator
Error from server (NotFound): configmaps "submariner-engine-lock" not found
$ kubectl delete configmap submariner-operator-lock -n submariner-operator
Error from server (NotFound): configmaps "submariner-operator-lock" not found
$ kubectl delete configmap submariner-lighthouse-coredns -n submariner-operator
Error from server (NotFound): configmaps "submariner-lighthouse-coredns" not found
```

---
title: "Uninstalling Submariner"
date: 2020-12-23T21:25:11+01:00
weight: 50
---

To properly uninstall Submariner from a cluster, follow the steps below:

{{% notice note %}}
Make sure KUBECONFIG for all participating clusters is exported and all participating clusters are accessible via kubectl.
{{% /notice %}}

1. Delete Submariner-related namespaces

   For each participating cluster, issue the following command:

   ```bash
   kubectl delete namespace submariner-operator
   ```

   For the Broker cluster, issue the following command:

   ```bash

>    kubectl delete namespace submariner-k8s-broker

Reviewer 5: With subctl-devel (0.9 and up), it is also required to delete submariner-operator namespace from the broker cluster:
see: https://github.com/submariner-io/submariner-operator/issues/203

Author: Correct. Added that section.

---
title: "Uninstalling Submariner"
date: 2020-12-23T21:25:11+01:00

> weight: 20

Reviewer 2: ```suggestion
weight: 50
```

Reviewer 2: This is to place this doc last under "Operations"

---

> To properly uninstall Submariner from a cluster, follow the steps below.

Reviewer 2: ```suggestion
To properly uninstall Submariner from a cluster, follow the steps below:
```

> 1. Delete Submariner related namespaces:

Reviewer 2: ```suggestion
1. Delete Submariner related namespaces
```

>    On data cluster,

Reviewer 2: ```suggestion
   On each participating cluster, issue the following command:
```

   ```bash
   kubectl delete namespace submariner-operator
   ```

>    On broker cluster,

Reviewer 2: ```suggestion
   On the Broker cluster, issue the following command:
```

   ```bash
   kubectl delete namespace submariner-k8s-broker
    ```

> 2. Deleting submariner CRDs in the cluster

Reviewer 2: ```suggestion
2. Delete the Submariner CRDs
```

> 

Reviewer 2: ```suggestion
On each participating cluster, issue the following command:
```

   ```bash
   for CRD in `kubectl get crds | grep -iE 'submariner|multicluster.x-k8s.io'| awk '{print $1}'`; do kubectl delete crd $CRD; done
   ```

3. Delete ClusterRoles and ClusterRoleBindings

> 

Reviewer 2: ```suggestion
On each participating cluster, issue the following command:
```

   ```bash

>    roles="submariner-operator submariner-operator-globalnet submariner-lighthouse"

Reviewer 6: We forgot this new component

```suggestion
   roles="submariner-operator submariner-operator-globalnet submariner-lighthouse submariner-networkplugin-syncer"
```

>    kubectl delete clusterrole,clusterrolebinding $roles

Reviewer 6: ```suggestion
   kubectl delete clusterrole,clusterrolebinding $roles --ignore-not-found
```
Just in case one of the objects does not exist try to avoid unnecessary errors.

   ```

> 4. Delete iptable chains, `SUBMARINER-INPUT` and `SUBMARINER-POSTROUTING`, that Submariner created.

Reviewer 2: ```suggestion
4. Remove Submariner's iptables chains
```

>    On all nodes, delete `SUBMARINER-INPUT`

Reviewer 2: ```suggestion
   On all nodes in each participating cluster, issue the following command:
```

   ```bash
   iptables --flush SUBMARINER-INPUT
   iptables -D INPUT $(iptables -L INPUT --line-numbers | grep SUBMARINER-INPUT | awk '{print $1}')
   iptables --delete-chain SUBMARINER-INPUT
   ```

>    Follow same commands for `SUBMARINER-POSTROUTING` chain.

Reviewer 2: Can you please add the right commands for this, so that a user can just copy-paste?

   If Globalnet is enabled in the setup, delete iptable chains `SUBMARINER-GN-INGRESS`, `SUBMARINER-GN-EGRESS` and `SUBMARINER-GN-MARK`

>    too following the same commands as above.

Reviewer 2: Ditto.

>    Note that KIND based setup will have just `SUBMARINER-INPUT` chain.

Reviewer 2: Style: use {{% notice note %}}

>    Note that KIND based setup will have just `SUBMARINER-INPUT` chain.

Reviewer 2: ```suggestion
   Note that kind based setup will have just `SUBMARINER-INPUT` chain.
```

> 5. Delete vx-submariner interface from all the nodes

Reviewer 2: ```suggestion
5. Delete `vx-submariner` interface
```

> 

Reviewer 2: On all nodes in each participating cluster, issue the following command:

   ```bash
   ip link delete vx-submariner
   ```

> 6. Remove submariner gateway labels

Reviewer 2: ```suggestion
6. Remove the Submariner gateway labels
```

> 

Reviewer 2: ```suggestion
On all nodes in each participating cluster, issue the following command:
```

---
title: "Uninstalling Submariner"
date: 2020-12-23T21:25:11+01:00
weight: 20
---

To properly uninstall Submariner from a cluster, follow the steps below.

1. Delete Submariner related namespaces:

>    On data cluster,

Reviewer 5: Is "Data cluster" the common name ? 
Shouldn't we say - on each cluster member (or other name we already use in Docs) ?

Author: data cluster is common IMO. And broker clusters. Thats the name we use, no?

Reviewer 5: @nyechiel ^

   ```bash
   kubectl delete namespace submariner-operator
   ```

   On broker cluster,

   ```bash
   kubectl delete namespace submariner-k8s-broker
    ```

2. Deleting submariner CRDs in the cluster

   ```bash
   for CRD in `kubectl get crds | grep -iE 'submariner|multicluster.x-k8s.io'| awk '{print $1}'`; do kubectl delete crd $CRD; done
   ```

3. Delete ClusterRoles and ClusterRoleBindings

>    ```bash

Reviewer 5: delete can get a list, e.g. 
```
roles="submariner-operator submariner-operator-globalnet submariner-lighthouse"
kubectl delete clusterrole,clusterrolebinding $roles
```

Author: Done
